Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp

```diff
@@ -635,29 +662,60 @@ RegisterObject::operator int()
                 reg.defaultTemplate = entry.defaultTemplate;
             }
         }
-        for (auto & creator_entry : entry.creatorMap)
+        for (auto& creator_entry : entry.creatorMap)
         {
-            const std::string & template_name = creator_entry.first;
+            const std::string& template_name = creator_entry.first;
             if (reg.creatorMap.find(template_name) != reg.creatorMap.end()) {
                 if (template_name.empty()) {
                     msg_warning("ObjectFactory") << "Class already registered: " << entry.className;
-                } else {
+                }
+                else {
                     msg_warning("ObjectFactory") << "Class already registered: " << entry.className << "<" << template_name << ">";
                 }
-            } else {
+            }
+            else {
                 reg.creatorMap.insert(creator_entry);
             }
         }
 
-        for (const auto & alias : entry.aliases)
+        for (const auto& alias : entry.aliases)
         {
             if (reg.aliases.find(alias) == reg.aliases.end())
             {
-                ObjectFactory::getInstance()->addAlias(alias,entry.className);
+                objectFactory->addAlias(alias, entry.className);
             }
         }
-        return 1;
+        return true;
+    }
+}
+
+typedef struct RegisterObjectsEntry
+{
+    inline static const char* symbol = "registerObjects";
+    typedef void (*FuncPtr) (sofa::core::ObjectFactory*);
+    FuncPtr func;
+    void operator()(sofa::core::ObjectFactory* data)
+    {
+        if (func) return func(data);
+    }
+    RegisterObjectsEntry() :func(nullptr) {}
+} RegisterObjectsEntry;
+
+bool ObjectFactory::registerObjectsFromPlugin(const sofa::helper::system::Plugin& plugin)
+{
+    sofa::helper::system::PluginManager& pluginManager = sofa::helper::system::PluginManager::getInstance();
```

Comment:
In the quest of removing the singletons, you could pass the pluginManager instance as a parameter